node/network/pov-distribution/src/lib.rs

```diff
@@ -540,50 +615,61 @@ impl PoVDistribution {
 	async fn run(
 		self,
 		mut ctx: impl SubsystemContext<Message = PoVDistributionMessage>,
-	) -> SubsystemResult<()> {
-		let mut state = State {
-			relay_parent_state: HashMap::new(),
-			peer_state: HashMap::new(),
-			our_view: View(Vec::new()),
-			metrics: self.metrics,
-		};
+	) -> crate::error::Result<()> {
+		let mut state = State::default();
+		state.metrics = self.metrics;
 
 		loop {
-			match ctx.recv().await? {
-				FromOverseer::Signal(signal) => if handle_signal(&mut state, &mut ctx, signal).await? {
-					return Ok(());
-				},
-				FromOverseer::Communication { msg } => match msg {
-					PoVDistributionMessage::FetchPoV(relay_parent, descriptor, response_sender) =>
-						handle_fetch(
-							&mut state,
-							&mut ctx,
-							relay_parent,
-							descriptor,
-							response_sender,
-						).await?,
-					PoVDistributionMessage::DistributePoV(relay_parent, descriptor, pov) =>
-						handle_distribute(
-							&mut state,
-							&mut ctx,
-							relay_parent,
-							descriptor,
-							pov,
-						).await?,
-					PoVDistributionMessage::NetworkBridgeUpdateV1(event) =>
-						handle_network_update(
-							&mut state,
-							&mut ctx,
-							event,
-						).await?,
-				},
+			while let Poll::Ready(Some((_relay_parent, _validator_id, peer_id))) =
+				futures::poll!(state.connection_requests.next()) {
+					if let Err(err) = handle_validator_connected(
+						&mut state,
+						peer_id,
+					).await {
+						warn!(
+							target: LOG_TARGET,
+							"Failed to handle validator connected: {:?}", err,
+						);
+					}
 			}
+
+			while let Poll::Ready(msg) = futures::poll!(ctx.recv()) {
+				match msg? {
+					FromOverseer::Signal(signal) => if handle_signal(&mut state, &mut ctx, signal).await? {
+						return Ok(());
+					},
+					FromOverseer::Communication { msg } => match msg {
+						PoVDistributionMessage::FetchPoV(relay_parent, descriptor, response_sender) =>
+							handle_fetch(
+								&mut state,
+								&mut ctx,
+								relay_parent,
+								descriptor,
+								response_sender,
+							).await?,
+						PoVDistributionMessage::DistributePoV(relay_parent, descriptor, pov) =>
+							handle_distribute(
+								&mut state,
+								&mut ctx,
+								relay_parent,
+								descriptor,
+								pov,
+							).await?,
+						PoVDistributionMessage::NetworkBridgeUpdateV1(event) =>
+							handle_network_update(
+								&mut state,
+								&mut ctx,
+								event,
+							).await?,
+					},
+				}
+			}
+
+			futures::pending!()
```

Comment:
In places like this we want to sort of manually re-create the logic of `select!` macro since using it in subsystem main loops usually runs into problems with borrowing. What we want is: poll everything that there is and when we've polled all things that are ready we need to tell the executor that we are pending until the next thing arrives. If you run the following code

```rust
use std::task::Poll;
use std::time::Duration;

use futures::{executor, StreamExt};

fn main() {
	executor::block_on(async move {
		let stream1 = futures::stream::repeat(()).then(|_| async move {
			futures_timer::Delay::new(Duration::from_secs(1)).await;
		});

		let stream2 = futures::stream::repeat(()).then(|_| async move {
			futures_timer::Delay::new(Duration::from_secs(1)).await;
		});

		futures::pin_mut!(stream1);
		futures::pin_mut!(stream2);

		loop {
			while let Poll::Ready(Some(_)) = futures::poll!(stream1.next()) {
				println!("stream1");
			}

			while let Poll::Ready(Some(_)) = futures::poll!(stream2.next()) {
				println!("stream2");
			}
		}
	});
}
```

you would find that without `futures::pending!()` it creates a busy loop since there is no place we back off.

Reply:
Where do we run here into borrowing issues and can not use `select!`?

Reply:
Thas is not always the case and here I changed the code to usage of `select_biased`.